<p style="font-size:30px;color:green;">Lesson 2: Знакомство с PANDAS</p>  

Tasks:
1. Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ';'. Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные. 
2. Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.
3. Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.
4. На сколько ночей в среднем бронируют отели разных типов?
5. Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?
6. Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?
7. Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов
8. Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?
9. Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?
10. Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.


<p style="font-size:20px;color:green;">Slove tasks</p>
<div id="toc"></div>
<script type="text/javascript" src="https://raw.github.com/kmahelona/ipython_notebook_goodies/master/ipython_notebook_toc.js">

### task 1 
**Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;.   
Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные.**

In [1]:
import pandas as pd

In [2]:
bookings = pd.read_csv('https://stepik.org/media/attachments/lesson/360344/bookings.csv', sep=';')

In [3]:
bookings.dtypes.value_counts()  # type of variables

int64      10
object     10
float64     1
dtype: int64

In [4]:
bookings_head = bookings.head(7)  # output 7 lines
bookings_head

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

### task 2

**Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.**

In [5]:
bookings = bookings.rename(columns= lambda c: c.lower().replace(' ', '_'))  # rename columns by lamda

In [6]:
bookings

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  adults  children  babies  meal country  \
0                          0  ...       2       0.0       0    BB     PRT   
1                          0  ...       2       0.0       0    BB     PRT   
2                          1  ...       1       0.0       0    BB     GBR   
3                          1  ...       1       0.0       0    BB     GBR   
4                          2  ...       2       0.0       0    BB     GBR   
...                      ...  ...     ...       ...     ...   ...     ...   
119385                     5  ...       2       0.0       0    BB     BEL   
119386                     5  ...       3       0.0       0    BB     FRA   
119387                     5  ...       2       0.0       0    BB     DEU   
119388                     5  ...       2       0.0       0    BB     GBR   
119389                     7  ...       2       0.0       0    HB     DEU   

       reserved_room_type assigned_room_type customer_type reservation_status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386         

In [7]:
bookings.columns  # list of all columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_full_date',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'stays_total_nights', 'adults', 'children',
       'babies', 'meal', 'country', 'reserved_room_type', 'assigned_room_type',
       'customer_type', 'reservation_status', 'reservation_status_date'],
      dtype='object')

### task 3
**Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.**

In [8]:
country_top5 = pd.DataFrame(
    bookings.query('is_canceled == 0')
    .country.value_counts()[:5]
    .reset_index()
    .rename(columns={'index':'country', 'country':'bookings'})
)  # interval for top5 contry, sort defolt in descending order

In [9]:
country_top5

country  bookings
0     PRT     21071
1     GBR      9676
2     FRA      8481
3     ESP      6391
4     DEU      6069

In [10]:
country_top5_bookings = country_top5.country.to_list()
print(*country_top5_bookings, sep=', ')

PRT, GBR, FRA, ESP, DEU


**Answer:** наибольшее число успешных бронирований в PRT, GBR, FRA, ESP, DEU (топ 5)

### task 4
**На сколько ночей в среднем бронируют отели разных типов?**

In [11]:
mean_hotel = (
    bookings.groupby("hotel", as_index=False)
    .agg({"stays_total_nights" : "mean"})
    .rename(columns={'stays_total_nights' : 'mean_night'})
    .round()
)

mean_hotel

hotel  mean_night
0    City Hotel         3.0
1  Resort Hotel         4.0

**Answer:** в отеле City Hotel в среднем бронируют 3 ночи, в отеле Resort Hotel в среднем бронируют 4 ночи

### task 5
**Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?**

In [12]:
room_type_non2 = bookings.query('reserved_room_type != assigned_room_type')  # all, but not 'assigned_room_type'

In [13]:
room_type_non2.shape[0] # amount lines = amount cases

14917

**Answer:** 14917 наблюдений

### task 6
**Проанализируйте даты запланированного прибытия. На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?**

In [14]:
arrival_date_2016 = pd.DataFrame(
    bookings.query("arrival_date_year == 2016")
    .value_counts('arrival_date_month')
    .reset_index()
    .rename(columns={0:'bookings'})
)  # booking in 2016 year every month

arrival_date_2016

arrival_date_month  bookings
0             October      6203
1                 May      5478
2               April      5428
3           September      5394
4                June      5292
5              August      5063
6               March      4824
7                July      4572
8            November      4454
9            February      3891
10           December      3860
11            January      2248

In [15]:
max_arrival_2016 = arrival_date_2016.bookings.max()  # month with maximum arrival in 20216
arrival_date_2016.query('bookings == @max_arrival_2016')

arrival_date_month  bookings
0            October      6203

In [16]:
arrival_date_2017 = pd.DataFrame(
    bookings.query("arrival_date_year == 2017")
    .value_counts('arrival_date_month')
    .reset_index()
    .rename(columns={0:'bookings'})
)  # booking in 2016 year every month

arrival_date_2017

arrival_date_month  bookings
0                May      6313
1              April      5661
2               June      5647
3               July      5313
4              March      4970
5             August      4925
6           February      4177
7            January      3681

In [17]:
max_arrival_2017 = arrival_date_2017.bookings.max()  # month with maximum arrival in 2017
arrival_date_2017.query('bookings == @max_arrival_2017')

arrival_date_month  bookings
0                May      6313

**Answer:** в 2016 самый популярный месяц - Октябрь, а в 2017 - Май. 

### task 7
**Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов**

In [18]:
canceled_top = pd.DataFrame(
    bookings.query("is_canceled == 1 and hotel == 'City Hotel'") 
    .groupby('arrival_date_year') 
    .arrival_date_month.value_counts()
)

canceled_top

arrival_date_month
arrival_date_year arrival_date_month                    
2015              September                         1543
                  October                           1321
                  August                            1232
                  July                               939
                  December                           668
                  November                           301
2016              October                           1947
                  June                              1720
                  September                         1567
                  April                             1539
                  May                               1436
                  November                          1360
                  August                            1247
                  March                             1108
                  December                          1072
                  July                              1043
                  February                           930
                  January                            438
2017              May                               2217
                  April                             1926
                  June                              1808
                  July                              1324
                  March                             1278
                  August                            1123
                  January                           1044
                  February                           971

In [19]:
max_arrival_2015 = canceled_top.query('arrival_date_year == 2015').arrival_date_month.max()
canceled_top.query('arrival_date_month == @max_arrival_2015')

arrival_date_month
arrival_date_year arrival_date_month                    
2015              September                         1543

In [20]:
max_arrival_2016 = canceled_top.query('arrival_date_year == 2016').arrival_date_month.max()
canceled_top.query('arrival_date_month == @max_arrival_2016')

arrival_date_month
arrival_date_year arrival_date_month                    
2016              October                           1947

In [21]:
max_arrival_2017 = canceled_top.query('arrival_date_year == 2017').arrival_date_month.max()
canceled_top.query('arrival_date_month == @max_arrival_2017')

arrival_date_month
arrival_date_year arrival_date_month                    
2017              May                               2217

**Answer:** самые частые отмены бронирования в отеле City Hotel:  
в 2015 - в Сентябре (1543 раза);  
в 2016 - в Октябре (1947 раз);  
в 2017 - в Мае (2217 раз)

### task 8
**Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?**

In [22]:
bookings[['adults', 'children', 'babies']].mean().max().round(2)

1.86

**Answer:** наибольшее среднее значение у adults

### task 9
**Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?**

In [23]:
bookings['total_kids']= bookings.children + bookings.babies

In [24]:
bookings.groupby('hotel').agg({'total_kids' : 'mean'}).round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

**Answer:** наибольшее среднее значение у отеля Resort Hotel 

### task 10
**Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.  
В качестве ответа укажите наибольший % оттока, округленный до 2 знаков после точки (то есть доля 0.24563 будет 24.56% и в ответ пойдёт 24.56)**

In [25]:
bookings['has_kids'] = bookings.total_kids > 0  # true if client have kids

In [26]:
bookings

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  babies  meal  country  reserved_room_type  \
0                          0  ...       0    BB      PRT                   C   
1                          0  ...       0    BB      PRT                   C   
2                          1  ...       0    BB      GBR                   A   
3                          1  ...       0    BB      GBR                   A   
4                          2  ...       0    BB      GBR                   A   
...                      ...  ...     ...   ...      ...                 ...   
119385                     5  ...       0    BB      BEL                   A   
119386                     5  ...       0    BB      FRA                   E   
119387                     5  ...       0    BB      DEU                   D   
119388                     5  ...       0    BB      GBR                   A   
119389                     7  ...       0    HB      DEU                   A   

       assigned_room_type customer_type reservation_status  \
0                       C     Transient          Check-Out   
1                       C     Transient          Check-Out   
2                       C     Transient          Check-Out   
3                       A     Transient          Check-Out   
4                       A     Transient          Check-Out   
...                   ...           ...                ...   
119385                  A     Transient          Check-Out   
119386                  E     Transient          Check-Out   
119387                  D     Transient          Check-Out   
119388 

In [27]:
bookings.query('is_canceled == 1 and has_kids == False').shape[0]  # count people who canceled the booking and not have kids

40965

In [28]:
bookings.query('is_canceled == 1 and has_kids == True').shape[0]  # count people who canceled the booking and have kids

3259

In [29]:
# Churn rate who canceled booking (how dependence on kids?)
no_kids_churn = bookings.query('is_canceled == 1 and has_kids == False').shape[0] / bookings.query('has_kids == False').shape[0]

no_kids_churn = round(no_kids_churn * 100, 2)  #precent 
no_kids_churn 

37.22

**Answer:** Показатель количества ушедших пользователей выше у тех, кто не имеет детей (37,22%)